src/rust/catnap/mod.rs

```diff
@@ -256,28 +180,17 @@ impl CatnapLibOS {
         debug_assert!((backlog > 0) && (backlog <= SOMAXCONN as usize));
 
         // Issue listen operation.
-        match self.qtable.borrow_mut().get_mut(&qd) {
-            Some(queue) => match queue.get_fd() {
-                Some(fd) => {
-                    // Create a listening socket.
-                    let listening_socket: Socket = {
-                        let bound_socket: &Socket = queue.get_socket();
-                        bound_socket.listen()?
-                    };
-
-                    if unsafe { libc::listen(fd, backlog as i32) } != 0 {
-                        let errno: libc::c_int = unsafe { *libc::__errno_location() };
-                        error!("failed to listen ({:?})", errno);
-                        return Err(Fail::new(errno, "operation failed"));
-                    }
-
-                    // Update socket.
-                    queue.set_socket(&listening_socket);
-                    Ok(())
-                },
-                None => unreachable!("CatnapQueue has invalid underlying file descriptor"),
+        let mut queue: CatnapQueue = self.get_queue(qd)?;
+        queue.prepare_listen()?;
```

Comment:
Same here, should we put the into queue.listen()?

Reply:
Yes, it could, but then the responsability for driving the state machine would be spread across the two layers. Eventually we want to place all of them in the queue layer, but this is not possible now.